---
title: "SNP Processing from UCEs for *Cerberus schneiderii*"
output:
  pdf_document: default
  html_document: default
---

## Processing of SNPs from Ultraconserved Elements (UCEs)

This script is to filter SNPs called from UCEs and investigating the effects of missing data using SNPfiltR.
Original script source: [Devon DeRaad](https://github.com/DevonDeRaad)
Date of Modification: March 2nd, 2023

  
Load packages
```
library(vcfR)
library(ggplot2)
library(adegenet)
library(SNPfiltR)
library(StAMPP)
```

# set working directory
setwd("~/Cerberus_Island-Biogeography/UCEs-SNPs-filter-missing-data-schneiderii-reduced-only-noMVZ253540/")

# read in vcf as vcfR
# this has been pre-filtered to remove SNPs near indels and SNPs mapping with low confidence
vcfR <- read.vcfR("~/Cerberus_Island-Biogeography/VCF-files-Cerberus/genotyped_X_schneiderii-reduced-noMVZ253540-samples_only_PASS_snp.vcf")
# 71,683 SNPs

# check out the details on the unfiltered vcf
vcfR.unfilt <- read.vcfR("~/Cerberus_Island-Biogeography/VCF-files-Cerberus/genotyped_X_schneiderii-reduced-noMVZ253540-samples_snps.vcf")
# 88,805 SNPs

# read in sample info csv
sample.info<-read.csv("sample.locs.csv")

# make sure sampling file matches vcf. The IDs need to be IDENTICAL to the names 
# you used while processing.
sample.info$ID %in% colnames(vcfR@gt)[-1]
colnames(vcfR@gt)[-1] %in% sample.info$ID

# if needed, subset sampling file to include only samples in vcf
# sample.info<-sample.info[sample.info$id %in% colnames(vcfR@gt)[-1],]

# reorder sampling file to match order of samples in vcf
sample.info<-sample.info[match(colnames(vcfR@gt)[-1], sample.info$ID),]
sample.info$ID == colnames(vcfR@gt)[-1]

# retain only biallelic SNPs
vcfR<-filter_biallelic(vcfR)
# 2031 SNPs, 0.028% of all input SNPs, contained more than 2 alleles, and were removed
# from the VCF

# execute allele balance filter
vcfR<-filter_allele_balance(vcfR)
# 73.67% of het genotypes (5.01% of all genotypes) fall outside of 0.25 - 0.75
# allele balance ratio and were converted to NA

# visualize and pick appropriate max depth cutoff
max_depth(vcfR)
# cutoff is not specified, exploratory visualization will be generated.
# dashed line indicates a mean depth across all SNPs of 36

# filter vcf by the max depth cutoff you chose
# looks like a decent number are >100, so let's pick 200
vcfR<-max_depth(vcfR, maxdepth = 200)
# maxdepth cutoff is specified, filtered vcfR object will be returned
# 0.57% of SNPs were above a mean depth of 200 and were removed from the vcf

# remove invariant SNPs
vcfR<-min_mac(vcfR, min.mac = 1)
# 1.8% of SNPs fell below a minor allele count of 1 and were removed from the VCF

# check vcfR to see how many SNPs we have left
vcfR
# 118 samples
# 4442 CHROMs
# 68,012 variants
# Object size: 122.6 Mb
# 4.563 percent missing data

#run function to visualize samples
miss<-missing_by_sample(vcfR=vcfR)

# every sample is missing less than 60% of SNPs, which is great.
# this means we should be able to retain all samples with a reasonable per-SNP 
# filtering threshold

# visualize missing data by SNP and the effect of various cutoffs on the 
# missingness of each sample
missing_by_snp(vcfR)
# Picking joint bandwidth of 0.0401
#    filt missingness     snps.retained
# 1  0.30  0.10910988         66198
# 2  0.50  0.09598002         64445
# 3  0.60  0.08745499         62930
# 4  0.65  0.08250466         61868
# 5  0.70  0.07701400         60501
# 6  0.75  0.07108728         58733
# 7  0.80  0.06377889         56042
# 8  0.85  0.05463047         51656
# 9  0.90  0.03758137         40819
# 10 0.95  0.02125539         27717
# 11 1.00  0.00000000          1370

# make popmap
popmap<-sample.info[sample.info$ID %in% colnames(vcfR@gt)[-1],c(5,1)]
# made popmap in excel; code on line 98 didn't work
popmap <- read.csv("popmap.csv")
colnames(popmap)<-c("id","pop")
popmap$pop<-as.factor(popmap$pop)

#assess missing data effects on clustering
assess_missing_data_pca(vcfR = vcfR, popmap = popmap, thresholds = c(.75,.85,1),
                        clustering = FALSE)
# cutoff is specified, filtered vcfR object will be returned
# 13.64% of SNPs fell below a completeness cutoff of 0.75 and were removed from the VCF
# cutoff is specified, filtered vcfR object will be returned
# cutoff is specified, filtered vcfR object will be returned
# 24.05% of SNPs fell below a completeness cutoff of 0.85 and were removed from the VCF
# cutoff is specified, filtered vcfR object will be returned
# 97.99% of SNPs fell below a completeness cutoff of 1 and were removed from the VCF

# in the tutorial, it was very weird that the sample used as a reference pops out
# but only above a certain filtering threshold. 

#may be a good idea to remove singletons (degraded samples) see how this affects the number of SNPs retained
vcfR.mac2<-min_mac(vcfR, min.mac = 2)
# 17.11% of SNPs fell below a minor allele count of 2 and were removed from the VCF

missing_by_snp(vcfR.mac2)
# Picking joint bandwidth of 0.041

#    filt  missingness    snps.retained
# 1  0.30  0.11812843         54663
# 2  0.50  0.10308559         52980
# 3  0.60  0.09340926         51542
# 4  0.65  0.08769207         50520
# 5  0.70  0.08139516         49217
# 6  0.75  0.07468226         47559
# 7  0.80  0.06647779         45071
# 8  0.85  0.05644252         41144
# 9  0.90  0.03850291         31989
# 10 0.95  0.02167492         21322
# 11 1.00  0.00000000           970

assess_missing_data_pca(vcfR = vcfR.mac2, popmap = popmap, thresholds = c(.70,.75,.80,1), clustering = FALSE)
# cutoff is specified, filtered vcfR object will be returned
# 12.69% of SNPs fell below a completeness cutoff of 0.7 and were removed from the VCF
# cutoff is specified, filtered vcfR object will be returned
# 15.63% of SNPs fell below a completeness cutoff of 0.75 and were removed from the VCF
# cutoff is specified, filtered vcfR object will be returned
# 20.05% of SNPs fell below a completeness cutoff of 0.8 and were removed from the VCF
# cutoff is specified, filtered vcfR object will be returned
# 98.28% of SNPs fell below a completeness cutoff of 1 and were removed from the VCF

# set 75% cutoff without minor allele filter
vcfRa<-missing_by_snp(vcfR, cutoff = .75)
# 13.64% of SNPs fell below a completeness cutoff of 0.75 and were removed from the VCF

# convert each to genlight
gena<-vcfR2genlight(vcfRa)

pop(gena)<-gena@ind.names
sample.div.75 <- stamppNeisD(gena, pop = FALSE)

# export for splitstree
# keep in mind that the names get truncated to 10 characters, so you will need to
# manually fix this or use gsub (below) to fix. Otherwise splitstree will split out
# error
stamppPhylip(distance.mat=sample.div.75, file="~/Cerberus_Island-Biogeography/UCEs-SNPs-filter-missing-data-schneiderii-reduced-only-noMVZ253540/splitstree/schneideri-reduced-noMVZ253540.75.splits.txt")

# for removing parts of names
# pop(gena) <- gsub("CS_MVZ", "", gena@ind.names)
#' @ gives a dataframe out of a list


# 75% completeness cutoff splitstree
# knitr is not working - to fix
knitr::include_graphics("~/Cerberus_Island-Biogeography/UCEs-SNPs-filter-missing-data-schneiderii-reduced-only-noMVZ253540/splitstree/schneideri-reduced-noMVZ253540.75.splits.txt.png")

#set 75% cutoff with mac = 2 filter
vcfRb<-missing_by_snp(vcfR.mac2, cutoff = .75)
# cutoff is specified, filtered vcfR object will be returned
# 15.63% of SNPs fell below a completeness cutoff of 0.75 and were removed from the VCF

#convert each to genlight
genb<-vcfR2genlight(vcfRb)

pop(genb)<-genb@ind.names
sample.div.mac.75 <- stamppNeisD(genb, pop = FALSE)
#export for splitstree
stamppPhylip(distance.mat=sample.div.mac.75, file="~/Cerberus_Island-Biogeography/UCEs-SNPs-filter-missing-data-schneiderii-reduced-only-noMVZ253540/splitstree/schneideri-reduced-noMVZ253540.75.mac.splits.txt")

#75% completeness cutoff with MAC splitstree
knitr::include_graphics(c("~/Cerberus_Island-Biogeography/UCEs-SNPs-filter-missing-data-schneiderii-reduced-only-noMVZ253540/splitstree/schneideri-reduced-noMVZ253540.75.mac.splits.txt.png"))

#set 85% cutoff with mac = 2 filter
vcfRc<-missing_by_snp(vcfR.mac2, cutoff = .85)
# cutoff is specified, filtered vcfR object will be returned
# 19.07% of SNPs fell below a completeness cutoff of 0.85 and were removed from the VCF

#convert each to genlight
genc<-vcfR2genlight(vcfRc)

pop(genc)<-genc@ind.names
sample.div.mac.85 <- stamppNeisD(genc, pop = FALSE)
# export for splitstree
stamppPhylip(distance.mat=sample.div.mac.85, file="/Documents/Cerberus_Island-Biogeography/UCEs-SNPs-filter-missing-data-schneiderii-reduced-only-noMVZ253540/splitstree/schneideri-reduced-noMVZ253540.85.mac.splits.txt")

#85% completeness cutoff with MAC splitstree
knitr::include_graphics(c("/Documents/Cerberus_Island-Biogeography/UCEs-SNPs-filter-missing-data-schneiderii-reduced-only-noMVZ253540/splitstree/schneideri-reduced-noMVZ253540.85.mac.splits.txt.png"))


# Interestingly, the two Borneo samples FMNH251594 and HKV31779 cluster with the other Sundaland samples rather than with the Philippines+Northern Sulawesi
# samples (N. Sulawesi = RMNH1391,1392 and MVZ253536) as is seen in the genomic phylogeny. The N. Sulawesi samples, though, still come out closer to the
# Philippines group rather than the other Wallacean samples (which is congruent with the phylogeny). 


# We will forge ahead with a 75% completeness and mac = 2 filtering scheme
# plot depth per snp and per sample
dp <- extract.gt(vcfRb, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)


# plot genotype quality per snp and per sample
gq <- extract.gt(vcfRb, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)


###We can use the convenient function 'write.vcf' from vcfR to export our filtered vcf file for downstream analyses
#write out vcf
vcfR::write.vcf(vcfRb, file = "~/Cerberus_Island-Biogeography/UCEs-SNPs-filter-missing-data-schneiderii-reduced-only-noMVZ253540/schneiderii-reduced-only-noMVZ253540.75.mac2.vcf.gz")
#perform linkage filtering to get a reduced vcf with only unlinked (one per UCE) SNPs
vcfR.thin<-distance_thin(vcfRb, min.distance = 10000)
# 4416 out of 47559 input SNPs were not located within 10000 
# base-pairs of another SNP and were retained despite filtering

vcfR.thin #retains 4,416 SNPs
# 118 samples
# 4416 CHROMs
# 4,416 variants
# Object size: 10.6 Mb
# 3.38 percent missing data

vcfR::write.vcf(vcfR.thin, file = "~/Cerberus_Island-Biogeography/UCEs-SNPs-filter-missing-data-schneiderii-reduced-only-noMVZ253540/schneiderii-reduced-only-noMVZ253540.75.mac2.unlinked.vcf.gz")
